### PR TITLE
hdf5: remove finalizers and document requirement for closing values

### LINKED ISSUE #23: hdf5: finalizers are sometimes called before they should
Hello @sbinet. It's me again :smile_cat: 

Using this library in production, I noticed some major issues that seem to be caused by the finalizers which are used.

Three out of four times, our application is crashing. I looked at the traces and found an interesting thing:
```
fatal error: unexpected signal during runtime execution
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x80 addr=0x0 pc=0xb762bb]

runtime stack:
runtime.throw(0xcd5fec, 0x2a)
        /usr/local/go/src/runtime/panic.go:616 +0x81
runtime.sigpanic()
        /usr/local/go/src/runtime/signal_unix.go:372 +0x28e

goroutine 4 [syscall]:
runtime.cgocall(0x88e8a0, 0xc42006b698, 0x0)
        /usr/local/go/src/runtime/cgocall.go:128 +0x64 fp=0xc42006b658 sp=0xc42006b620 pc=0x401f94
decode/vendor/gonum.org/v1/hdf5._Cfunc_H5Tclose(0x30000000000013e, 0x0)
        _cgo_gotypes.go:913 +0x4d fp=0xc42006b698 sp=0xc42006b658 pc=0x71391d
decode/vendor/gonum.org/v1/hdf5.(*Datatype).Close(0xc42022cbd0, 0xc42006b738, 0x0)
        /go/src/decode/vendor/gonum.org/v1/hdf5/h5t.go:153 +0x37 fp=0xc42006b6d0 sp=0xc42006b698 pc=0x717ac7
decode/vendor/gonum.org/v1/hdf5.(*Datatype).finalizer(0xc42022cbd0)
        /go/src/decode/vendor/gonum.org/v1/hdf5/h5t.go:138 +0x2f fp=0xc42006b728 sp=0xc42006b6d0 pc=0x7179ef
runtime.call32(0x0, 0xce3b48, 0xc420432000, 0x1000000010)
        /usr/local/go/src/runtime/asm_amd64.s:573 +0x3b fp=0xc42006b758 sp=0xc42006b728 pc=0x4535fb
runtime.runfinq()
        /usr/local/go/src/runtime/mfinal.go:222 +0x1eb fp=0xc42006b7e0 sp=0xc42006b758 pc=0x415edb
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc42006b7e8 sp=0xc42006b7e0 pc=0x455d81
created by runtime.createfing
        /usr/local/go/src/runtime/mfinal.go:156 +0x62
```

Looks like the `(Datatype).finalizer` tries to close an already closed Datatype... hmm... This happened right after the beginning of execution where the finalizer should actually not be called.

Looking at other traces, I noticed that the `Close()` on `Datatype` is not always the cause of crash, but a segmentation violation in regard to that `Datatype` is.

Having extensive logging in place that also spits out a log whenever we call `Close()` on the `Datatype` ourselves, I noticed that this actually never happened in those cases. So something else must've closed it before.

So I went to our `vendor` folder, `h5t.go`, removed the finalizer related code and *BOOOM*. The error does not occur any longer. Our application calls `Close()` when it is actually done writing and everything is fine.

As we both know (latest after the article that you linked recently), finalizers are not optimal. Actually, they seem to be causing issues in regard to this library. Therefore, I would suggest to remove them all over the place and would also take care of it in case you agree.

IMHO everybody that uses this library should actually take care of the `Close()` calls themselves.

For the record, I am running `Go 1.10` and `libhdf5 1.10.1`.

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Fixes #23.

### CODE REVIEW

File: h5f.go [hunk 1]
```diff
@@ -66,7 +56,8 @@ func CreateFile(name string, flags int) (*File, error) {
 	return newFile(hid), nil
 }
 
-// Opens an existing HDF5 file.
+// Open opens and returns an an existing HDF5 file. The returned
```
Comment (TacoVox) [inline h5f.go]: Just a general thought: should we maybe stay consistent with the method/function comments and start them with method/function name. Right now it seems to be the case for half of the ones below.

File: h5s.go [hunk 2]
```diff
@@ -73,6 +64,7 @@ func h5sclose(id C.hid_t) C.herr_t {
 }
 
 // CreateSimpleDataspace creates a new simple dataspace and opens it for access.
+//  The returned dataspace must be closed by the user when it is no longer needed.
```
Comment (TacoVox) [inline h5s.go]: Remove one white space :sunglasses:

File: h5f.go (hunk 1 above)
Comment (kortschak) [reply h5f.go]: Yeah, I missed some. The intention was to do "Func does ....". We can fix more later. There's other lint I want to remove elsewhere, so I'm fine with leaving this and repolishing.

File: h5s.go (hunk 2 above)
Comment (kortschak) [reply h5s.go]: This one needs to be fixed. Thanks.

File: h5f.go (hunk 1 above)
Comment (TacoVox) [reply h5f.go]: SGTM

File: h5t.go [hunk 3]
```diff
@@ -109,13 +109,12 @@ func OpenDatatype(c CommonFG, name string, tapl_id int) (*Datatype, error) {
...
 func NewDatatype(id C.hid_t) *Datatype {
 	t := &Datatype{Identifier{id}}
-	runtime.SetFinalizer(t, (*Datatype).finalizer)
 	return t
 }
 
-// CreateDatatype creates a new datatype.
-// class must be T_COMPUND, T_OPAQUE, T_ENUM or T_STRING.
-// size is the size of the new datatype in bytes.
+// CreateDatatype creates a new datatype. The value of class must be T_COMPUND,
```
Comment (sbinet) [inline h5t.go]: actually, it's `s/T_COMPUND/T_COMPOUND/` :)

File: h5t.go [hunk 4]
```diff
@@ -264,15 +255,15 @@ type CompoundType struct {
 	Datatype
 }
 
-// NewCompoundType creates a new CompoundType.
-// size is the size in bytes of the compound datatype.
+// NewCompoundType creates a new CompoundType. The/ size is the size in bytes of
```
Comment (sbinet) [inline h5t.go]: `s|The/|The|`

File: h5t.go (hunk 3 above)
Comment (kortschak) [reply h5t.go]: Fixed

File: h5t.go (hunk 4 above)
Comment (kortschak) [reply h5t.go]: Fixed

**Pull request merged**